File: src/check_comlogic.c
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <unistd.h>
#include "pr_utils.h" 
#include "cce_1_def.h" 

/*
		gcc -o check_comlogic check_comlogic.c -lpigpio -lpthread
	
*/

int My_logic_function ( int* inp , int insize, int*  outp, int outsize  ) {
	if ( insize > 3 ) return -1; 
	if ( outsize > 2 ) return -2; 
	// make sure input is only 1 or 0 
	for ( int inpcnt=0; inpcnt < insize ; inpcnt++) {
			inp[inpcnt] = inp[inpcnt] & 1;
	}
	outp[0] = inp[0] & inp[1]; // bit wize or 
	outp[1] = ! (outp[0]  | inp[2]) ; 
	return 0;
} 
	


int main(int argc, char *argv[]){
// --- this is setup the hardware for the test 
	int err= init_cce_1();
	if (err) {
		printf("initialization of hardware failed with err %d \n\r", err);
		return err; 
	}
	err=set_muxout( 'A' );
	if (err) {
		printf("can not set the output mux correctly for this test  err %d \n\r", err);
		return err; 
	}
// define test specific input / outputs
	// now define my inputs 
	const int nr_inputs=3;
	unsigned  inputs[]= {D0,D1,D2} ;
	if (sizeof(inputs) / sizeof(inputs[0]) < nr_inputs) { 
		printf("input pin array badly defined \n\r");
		return -nr_inputs ;
	}
// these drive the inputs of the FPGA so have to defined as outputs 
	for ( int inpcnt=0; inpcnt < nr_inputs; inpcnt++) {
		err= gpioSetMode( inputs[0],PI_OUTPUT);
		if( err) { printf("can not set gpio %d to output with err %d \n\r",inputs[inpcnt], err); 
			return err;
		}
	}
   
   // the output are defined in the general hardware setup 		
   // we need only 1 output but we use 2 , msb is always 1 in the simulation so we get errors in the test 
	const int nr_outputs=2; 
	
// -- here start the test 	
	int nr_inpcomb = 1;nr_inpcomb = nr_inpcomb<<nr_inputs ;  // 2 to the power of nr_inputs 
	int bin_outp_arry[nr_outputs];
	int simulations_inputs[nr_inputs];
	int simulations_outputs[nr_outputs];
	err=0;	
	// itterate  over all combinations 
    for (int  inpcomb=0 ; inpcomb < nr_inpcomb ; inpcomb++) {
		set_outputs( inputs, nr_inputs , inpcomb) ;
		// convert to bin array for the simulation 
		decvalue2binar(inpcomb ,simulations_inputs  ,nr_inputs );
		// simulate 
		My_logic_function ( simulations_inputs , nr_inputs,simulations_outputs, nr_outputs  );
		// get the outputs of the circuit 
		get_led_status_bin( bin_outp_arry, 2 ); 
		// now we print all in a format so we can easy check it with the LED status 
		printf ("input:  ");
		print_bin_arry_status (simulations_inputs , nr_inputs );
		printf(" output : ");
		print_bin_arry_status (bin_outp_arry, nr_outputs);
		printf(" simout : ");
		print_bin_arry_status (simulations_outputs, nr_outputs);
		printf("\n\r");
		//give some  time to check the LED's on the FPGA board 
		sleep(2);
		
		
	}
	 gpioTerminate();
	return 0;  // will not be reached. 
}

File: python/check_comlogic.py
import time
from pr_utils import *
from upctrl_pinfunctions import *


def my_logic_function(inp, outp):
    insize = len(inp)
    outsize = len(outp)
    if insize > 3:
        return -1
    if outsize > 2:
        return -2
    for inpcnt in range(insize):
        inp[inpcnt] = bool(inp[inpcnt] & 1)
    outp[0] = int(inp[0] & inp[1])
    outp[1] = int(not (outp[0] | inp[2]))
    return 0

def main():
    pi,err = init_cce_1()
    if err:
        print("initialization of hardware failed with err", err)
        return err
    set_muxout('A')  # set the routing of the signals on the FPGA board
    # define test specific input / outputs
    nr_inputs = 3
    inputs = [D0, D1, D2]
    #these drive the inputs of the FPGA so have to defined as outputs 
    for Dx in inputs :
        pi.set_mode(Dx, pigpio.OUTPUT)
    nr_outputs = 2
    #the output are defined in the general hardware setup 		
    simulations_inputs = [0,0,0]
    simulations_outputs = [0,0] 
    err = 0
    nr_inpcomb = 8 # 2^3
    for inpcomb in range(nr_inpcomb):
        set_outputs(inputs, nr_inputs, inpcomb)
        err,simulations_inputs=decvalue2binar(inpcomb,  nr_inputs)
        my_logic_function(simulations_inputs, simulations_outputs)
        bin_outp_arry=get_led_status_bin( 2)
        print("input: ", end='')
        print_bin_arry_status(simulations_inputs, nr_inputs)
        print("output: ", end='')
        print_bin_arry_status(bin_outp_arry, nr_outputs)
        print("simout: ", end='')
        print_bin_arry_status(simulations_outputs, nr_outputs)
        print()
        time.sleep(2)
    # gpioTerminate() is omitted in Python translation

if __name__ == "__main__":
    main()


